SU2_CFD/src/solvers/CNEMOEulerSolver.cpp

```diff
@@ -815,45 +815,70 @@ void CNEMOEulerSolver::Source_Residual(CGeometry *geometry, CSolver **solver_con
 
   unsigned short iVar, jVar;
   unsigned long iPoint;
-  unsigned long eAxi_local, eChm_local, eVib_local;
-  unsigned long eAxi_global, eChm_global, eVib_global;
 
   /*--- Assign booleans ---*/
   bool err        = false;
   bool implicit   = (config->GetKind_TimeIntScheme_Flow() == EULER_IMPLICIT);
   bool frozen     = config->GetFrozen();
   bool monoatomic = config->GetMonoatomic();
+  bool axisymm    = config->GetAxisymmetric();
   bool viscous    = config->GetViscous();
   bool rans       = (config->GetKind_Turb_Model() != NONE);
 
   CNumerics* numerics = numerics_container[SOURCE_FIRST_TERM];
 
   /*--- Initialize the error counter ---*/
-  eAxi_local = 0;
-  eChm_local = 0;
-  eVib_local = 0;
+  unsigned long eAxi_local = 0;
+  unsigned long eChm_local = 0;
+  unsigned long eVib_local = 0;
 
   /*--- Initialize the source residual to zero ---*/
   for (iVar = 0; iVar < nVar; iVar++) Residual[iVar] = 0.0;
 
+  /*--- Preprocess viscous axisymm variables (if necessary) ---*/
+  if (axisymm && viscous) {
+
+    for (iPoint = 0; iPoint < nPoint; iPoint++) {
+      su2double yCoord          = geometry->nodes->GetCoord(iPoint, 1);
+      su2double yVelocity       = nodes->GetVelocity(iPoint,1);
+      su2double xVelocity       = nodes->GetVelocity(iPoint,0);
+      su2double Total_Viscosity = nodes->GetLaminarViscosity(iPoint) + nodes->GetEddyViscosity(iPoint);
+
+      if (yCoord > EPS){
+        su2double nu_v_on_y = Total_Viscosity*yVelocity/yCoord;
+        nodes->SetAuxVar(iPoint, 0, nu_v_on_y);
+        nodes->SetAuxVar(iPoint, 1, nu_v_on_y*yVelocity);
+        nodes->SetAuxVar(iPoint, 2, nu_v_on_y*xVelocity);
+      }
+   }
+
+   /*--- Compute the auxiliary variable gradient with GG or WLS. ---*/
+   if (config->GetKind_Gradient_Method() == GREEN_GAUSS) {
+     SetAuxVar_Gradient_GG(geometry, config);
+   }
+   if (config->GetKind_Gradient_Method() == WEIGHTED_LEAST_SQUARES) {
+     SetAuxVar_Gradient_LS(geometry, config);
+   }
```

Comment:
Move this to CFVMFlowSolverBase.inl and share it between CEulerSolver and CNEMOEulerSolver please (the bug is in CEuler solver, not here)

Reply:
Yes, the logic is more or less: if the method does not have to be a template then it is declared in the .hpp and implemented in .inl of CFVMSolverBase.
If it has to be a template, or you want it to be "inlineable", then it has to go in the .hpp.
But both files are headers, the only thing that changes is the "visibility", e.g. CEulerSolver includes the hpp but not the inl.
At the moment the inl is included by the NS solvers, at the top of the associated .cpp's you'll see
```c++
/*--- Explicit instantiation of the parent class of CIncEulerSolver,
 *    to spread the compilation over two cpp files. ---*/
template class CFVMFlowSolverBase<CIncEulerVariable, ENUM_REGIME::INCOMPRESSIBLE>;
```
That also means that if you change something in the .inl only the NS solvers recompile, but not the Euler solvers (which sounded like a good idea because e.g. CEulerSolver is a reasonably large file).

The incompressible version looks different enough to be either its own method or left as is.
My rule of thumb for such master versions is not to have physics-based conditionals in there (e.g. "if compressible"), if necessary, that type of detail should be provided from the outside, and not already present in the body of the "master" version.
In other words, we want those versions to be generic in the sense they allow externally provided specialization, not in the sense they already try to cover all possible options.
I hope this makes some sense.

Reply:
Done!